core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -1162,6 +1162,13 @@ class Partition(val topicPartition: TopicPartition,
     localLog.fetchOffsetSnapshot
   }
 
+  def hasDivergingEpoch(currentLeaderEpoch: Optional[Integer],
+                        lastFetchedEpoch: Int,
+                        fetchOffset: Long): Boolean = {
+    val epochEndOffset = lastOffsetForLeaderEpoch(currentLeaderEpoch, lastFetchedEpoch, fetchOnlyFromLeader = false)
```

Comment:
Do we need to check the error? Or are you relying on the check below failing if `UNDEFINED_EPOCH` is returned?

Reply:
@hachikuji Thanks for the review. I took the check from `Partition.readRecords`, but we throw exceptions there to return appropriate errors. I was thinking we would return true here for undefined epochs because of the check below and that would go through the other code path to return the appropriate errors or diverging epoch. Do you think we should do the same error handling hchecks ere as in readRecords to make the flow more obvious?

Reply:
As discussed offline, moved the check to DelayedFetch.